### PR TITLE
NEXT-24962 - Add IPv6 CIDR suggestions to maintenance configuration

### PR DESCRIPTION
**Pull request by JoshuaBehrens**
### 1. Why is this change necessary?

IPv6 exists.

### 2. What does this change do, exactly?

Allow CIDR configuration in the administration. The IPUtils from Symfony already support CIDR.
So now you can just enter IPv6 and IPv4 CIDR and you get suggestions for common IPv6 CIDRs based on your IPv6 address so you can easily get all of your mates aboard. Arrrr!

Previously:

image
image

With this PR:

image
image
image

### 3. Describe each step to reproduce the issue or behaviour.

1. Surf the shop with IPv6 but be used to IPv4
2. Start maintenance mode
3. Select "You" in the configuation suggestion
4. Share the knowledge, that "we are now excluded from maintenance mode"
5. Everybody response "I still get maintenance mode"

### 4. Please link to the relevant issues (if any).

https://issues.shopware.com/issues/NEXT-24962

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### <samp>🤖 Generated by Copilot at 418ebb8</samp>

This pull request adds support for IPv6 addresses and CIDR masks to the trusted proxies configuration in the Administration module. It updates the `KnownIpsCollector` class, the `sw-multi-tag-ip-select` component, the string utils and the translation files. It also adds a new changelog entry and validation logic for the maintenance IP configuration.

### <samp>🤖 Generated by Copilot at 418ebb8</samp>

*  Add support for IPv6 CIDR masks to the known IP addresses configuration ([link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-47d11c5cc7ad3fe665874ed44660e314c5356a262798a76692fd62379ccaf015R1-R17))
*  Update the `KnownIpsCollector` class to return IPv6 CIDR masks based on the client IP ([link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-214fff05b7a26436808fd4cb0b4d3eaae352bbfaa1fcf0d9cce1b51d95c64fecL15-R19), [link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-214fff05b7a26436808fd4cb0b4d3eaae352bbfaa1fcf0d9cce1b51d95c64fecL22-R53))
*  Add a new prop `errorCode` to the `sw-multi-tag-ip-select` component to customize the error message ([link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-e1f028eec6d4af49d30cf6e754e1ead0dabf60633802491dbdf34329a6c3ea7eR35-R42))
*  Modify the `sw-multi-tag-ip-select` component to validate both IP addresses and CIDR masks using the `validate` method ([link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-e1f028eec6d4af49d30cf6e754e1ead0dabf60633802491dbdf34329a6c3ea7eL41-R53))
*  Add a new snippet `SHOPWARE_INVALID_IP_CIDR` to the German and English translation files for the error message ([link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-6829d4ef8031fc00da5e01f7a12447fb257951c92db10c54fb395e9d46ff8843R11), [link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-276bd8f8b050d50fad6d16c728fc17f798e275bcf2636fd39a0128d9e5d203c5R11))
*  Add new snippets `youIPv6Block56` and `youIPv6Block64` to the German and English translation files for the suggested IPv6 CIDR masks ([link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-6829d4ef8031fc00da5e01f7a12447fb257951c92db10c54fb395e9d46ff8843L541-R544), [link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-276bd8f8b050d50fad6d16c728fc17f798e275bcf2636fd39a0128d9e5d203c5L541-R544))
*  Expose the `isValidCidr` function from the `string.utils.ts` file to the `util.service.ts` file ([link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-b58bd49a7e6e72cdc6045d990efacb4494ed808554a86b6cce904fd8f403f630R88), [link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-9331eb6b4a9fba78a1339725dd7681525ad3df90b8f9fd0611c7a39ec5d32dd3R22))
*  Implement the `isValidCidr` function in the `string.utils.ts` file using a regular expression ([link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-9331eb6b4a9fba78a1339725dd7681525ad3df90b8f9fd0611c7a39ec5d32dd3R50-R55))
*  Add the `validateMaintenanceIpCidr` method to the `sw-sales-channel-detail-base` component to validate the maintenance IP configuration ([link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-5ba1d73acccd6c9cf0025d855e43bed595b3b8a7afa1e143c03fada92cb36709R724-R727))
*  Add the `validate` and `error-code` attributes to the `sw-multi-tag-ip-select` component in the `sw-sales-channel-detail-base` template ([link](https://github.com/shopware/platform/pull/3296/files?diff=unified&w=0#diff-34fe5ea5435e329c49fbdd3349bf5da76001f54a41a77931735442387dd4fe80R810-R811))

### CODE REVIEW

File: src/Administration/Framework/Routing/KnownIps/KnownIpsCollector.php [hunk 1]
```diff
@@ -12,18 +12,43 @@ class KnownIpsCollector implements KnownIpsCollectorInterface
...
-     *         '127.0.0.1' => 'global.sw-multi-tag-ip-select.knownIps.you'
+     *         '127.0.0.1' => 'global.sw-multi-tag-ip-select.knownIps.you',
+     *         // or
+     *         '2001:0db8:0123:4567:89ab:cdef:1234:5678' => 'global.sw-multi-tag-ip-select.knownIps.you',
+     *         '2001:0db8:0123:4567::/64' => 'global.sw-multi-tag-ip-select.knownIps.youIPv6Block64',
+     *         '2001:0db8:0123:4500::/56' => 'global.sw-multi-tag-ip-select.knownIps.youIPv6Block56'
      *     ]
      * </code>
      */
     public function collectIps(Request $request): array
```
Comment (shyim) [inline src/Administration/Framework/Routing/KnownIps/KnownIpsCollector.php:23]: Can you add a unit test to the code?

File: src/Administration/Framework/Routing/KnownIps/KnownIpsCollector.php [hunk 2]
```diff
@@ -12,18 +12,43 @@ class KnownIpsCollector implements KnownIpsCollectorInterface
...
-        if ($request->getClientIp()) {
-            $result[$request->getClientIp()] = 'global.sw-multi-tag-ip-select.knownIps.you';
+        if ($clientIp) {
+            $result[$clientIp] = 'global.sw-multi-tag-ip-select.knownIps.you';
+
+            $isIpV6 = \str_contains($clientIp, ':') && \filter_var($clientIp, \FILTER_VALIDATE_IP, \FILTER_FLAG_IPV6);
+
+            if ($isIpV6 && $this->canPhpIpv6()) {
+                // replace last half (and slightly more for /56)
+                $cidr64 = \inet_ntop(\hex2bin(\substr(\bin2hex(\inet_pton($clientIp)), 0, 16) . \str_repeat('0', 16))) . '/64';
```
Comment (shyim) [inline src/Administration/Framework/Routing/KnownIps/KnownIpsCollector.php:35]: Can you fix the PHPStan errors here?

**Comment by JoshuaBehrens** [comment]
I rebased and added a test but I am not sure where to spot the phpstan error

File: tests/unit/Administration/Framework/Routing/KnownIps/KnownIpsCollectorTest.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+use PHPUnit\Framework\TestCase;
+use Shopware\Administration\Framework\Routing\KnownIps\KnownIpsCollector;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * @internal
+ */
+final class KnownIpsCollectorTest extends TestCase
+{
+    public function testDeleteWillRemoveUserConfigs(): void
```
Comment (shyim) [inline tests/unit/Administration/Framework/Routing/KnownIps/KnownIpsCollectorTest.php]: can you fix the method name, I think you copied it but didn't changed the method name

File: tests/unit/Administration/Framework/Routing/KnownIps/KnownIpsCollectorTest.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Tests\Unit\Administration\Framework\Routing\KnownIps;
+
+use PHPUnit\Framework\TestCase;
+use Shopware\Administration\Framework\Routing\KnownIps\KnownIpsCollector;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * @internal
```
Comment (shyim) [inline tests/unit/Administration/Framework/Routing/KnownIps/KnownIpsCollectorTest.php:11]: as this is an unit test, we need here an `@covers`

File: tests/unit/Administration/Framework/Routing/KnownIps/KnownIpsCollectorTest.php (hunk 3 above)
Comment (JoshuaBehrens) [reply tests/unit/Administration/Framework/Routing/KnownIps/KnownIpsCollectorTest.php]: Maybe :DDD ty for noticing this

**Comment by JoshuaBehrens** [comment]
I rebased and amend regarding the comments :)

File: tests/unit/Administration/Framework/Routing/KnownIps/KnownIpsCollectorTest.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Tests\Unit\Administration\Framework\Routing\KnownIps;
+
+use PHPUnit\Framework\TestCase;
+use Shopware\Administration\Framework\Routing\KnownIps\KnownIpsCollector;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * @covers \Shopware\Administration\Framework\Routing\KnownIps\KnownIpsCollector
```
Comment (larskemper) [inline tests/unit/Administration/Framework/Routing/KnownIps/KnownIpsCollectorTest.php]: The `@covers` annotation is not in use anymore :sweat_smile:. Please use the class attribute: `#[CoversClass(KnownIpsCollectorTest::class)]`

File: src/Administration/Resources/app/administration/src/core/service/utils/string.utils.ts [hunk 6]
```diff
@@ -46,3 +47,9 @@ function isValidIp(ip: string): boolean {
...
 }
+
+function isValidCidr(cidr: string): boolean {
```
Comment (larskemper) [inline src/Administration/Resources/app/administration/src/core/service/utils/string.utils.ts:51]: Please add a test case

File: src/Administration/Resources/app/administration/src/module/sw-sales-channel/view/sw-sales-channel-detail-base/index.js [hunk 7]
```diff
@@ -725,5 +725,9 @@ export default {
         isFavorite() {
             return this.salesChannelFavoritesService.isFavorite(this.salesChannel.id);
         },
+
+        validateMaintenanceIpCidr(term) {
```
Comment (larskemper) [inline src/Administration/Resources/app/administration/src/module/sw-sales-channel/view/sw-sales-channel-detail-base/index.js:726]: Please add a test case

**Review by larskemper — CHANGES_REQUESTED** [review]

File: src/Administration/Resources/app/administration/src/module/sw-sales-channel/view/sw-sales-channel-detail-base/index.js (hunk 7 above)
Comment (JoshuaBehrens) [reply src/Administration/Resources/app/administration/src/module/sw-sales-channel/view/sw-sales-channel-detail-base/index.js:726]: Is it easier now to run vue tests?

Comment (larskemper) [reply src/Administration/Resources/app/administration/src/module/sw-sales-channel/view/sw-sales-channel-detail-base/index.js:726]: If you are referring to the transition from Vue 2 to Vue 3, then yes. However, in a broader context, there haven't been significant updates in this regard. Specifically in your situation, the task should be relatively straightforward. The `string.utils.ts` file is not a component (take a look at something similar, like [types.utils.spec.js](https://github.com/shopware/shopware/blob/trunk/src/Administration/Resources/app/administration/src/core/service/utils/types.utils.spec.js)), and the `sw-sales-channel-detail` component already has a good test.

**Comment by JoshuaBehrens** [comment]
I rebased, changed code regarding comments and added tests :)

**Comment by mitelg** [comment]
thanks for your contribution @JoshuaBehrens 👍🎉💙

**Comment by JoshuaBehrens** [comment]
Thank you for merging @mitelg  :)

**Pull request closed without merging**